### ISSUE TITLE
Adding a non-HA user

### ISSUE THREAD

**Issue by SaturnusDJ** [gcobb321/icloud3#65]
Hey Gary,

Making a new issue as that really makes more sense now.

My goal: Adding a friend to the HA map. So this is not myself and this friend does not have an account nor access to HA.
I can see the friend in the 'Find My' app with regular updates.

Running 2.2.0-rc11b.

Device_tracker:
```
  - platform: icloud3
    username: bla@icloud.com
    password: 'realpass'
    tracking_method: fmf
    config_ic3_file_name: custom_components/icloud3/config_ic3.yaml
```
The above account information is my iCloud login.

config_ic3.yaml:
```
track_devices:
  - Friend > friend@icloud.com
```

Here `Friend` equals the name shown in the 'Find My' app which is the contact name as in my contacts. Also made the friend change their phone name to that value.
`friend@icloud.com` equals the e-mail address that is saved to this contact and also used by them for the 'Find My' app.

Log:
```
2020-08-20 16:02:48 WARNING (MainThread) [homeassistant.loader] You are using a custom integration for icloud3 which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant.
2020-08-20 16:02:54 ERROR (SyncWorker_7) [custom_components.icloud3.device_tracker] Two-Step Authentication (2SA) Required for Account bla@icloud.com
Traceback (most recent call last):
  File "/config/custom_components/icloud3/device_tracker.py", line 6067, in _setup_tracked_devices_for_fmf
    self._log_level_debug_rawdata("iCloud FmF Raw Data - (device_obj.data)", device_obj.data)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 717, in data
    self.refresh_client()
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 688, in refresh_client
    req = self.session.post(self._friend_endpoint, data=mock_payload, params=params)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 578, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 153, in request
    self._raise_error(code, reason)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 168, in _raise_error
    raise PyiCloud2SARequiredException(self.service.user["apple_id"])
custom_components.icloud3.pyicloud_ic3.PyiCloud2SARequiredException: Two-Step Authentication (2SA) Required for Account bla@icloud.com
2020-08-20 16:02:54 ERROR (SyncWorker_7) [custom_components.icloud3.device_tracker]  iCloud3 Setup Error >  > The iOS App device_tracker entity was not found in the Entity Registry.. Device_tracker entities (suffixes) found -
```

Other things that happen:
* Getting asked to allow the sign in on iPhone every restart of HA.
* No text message or notification in HA.
* The device_tracker entry of the friend is shown as `Home` while being far away from 'Home'.

I am not sure if this is a bug or if I am making a mistake. Been in the docs (https://gcobb321.github.io/icloud3_docs/#/chapters/2-config-parms) for some hours now.

___

When trying it for myself instead of friend, 

```
 iOS App Entity device_tracker.me_2  does not contain location attributes (latitude, longitude).
```
While it does.

**Comment by gcobb321** [comment]
Maybe both. I think there is a bug not catching a 2SA needed Authentication in the fmf routine and am looking at it now. This friend will need an iCloud account but not an HA account. The friend will need the iOS App loaded on their phone and will have to log into your HA account to get zone enter/exit and location triggers.

I'll keep you posted.

**Comment by gcobb321** [comment]
With the error:
1. Do you have log_level: debug+rawdata in the config file. If so, take off the rawdata and see if it gets the same error
2. Have you authenticated the bla@icloud.com account that starts with a message in the HA Notifications area?
3. Try and get yourself up and running successfully. Then once you are OK, add your friend.

The error is generated from an uncaught error message that your account needs to be authenticated. Which is kinda complicated  and something I can not reproduce since i am slready authenticated.

Send your real email address to geekstergary@gmail.com. I'll probably send you a special version with some test code in it tomorrow. I have been working with several others in the same manner in the past.

**Comment by SaturnusDJ** [comment]
1. It is at default.
2. In all the hours of testing, not one notification came up in HA about this software.
3. Alright, will mail tomorrow.

**Comment by gcobb321** [comment]
So it seems to be ok when it’s just tracking you?
But the Errors are created when you add your friend?

**Comment by gcobb321** [comment]
I think it's an iCloud authentication/verification code issue that needs to be done before the fmf & famshr devices can be set up. I've uploaded Release Candidate 11c zip file to the Release Candidate directory  that has code that I think may correct the problem. It should respond to the Verification Request when iCloud3 starts. It will put some info in the Event Log that it is needed and use the iOS App tracking method until it has been done. After successfully completing the verification, it will reset to use fmf or famshr.

I have had some problems where the Trusted Device ID selection and Verification Code Entry forms would not go away. This was caused by the selected item being rejected by Apple because it was incorrect even though it was OK. That may be some type of other error or I just banged on Apple to many times in a short period of time. Restarting HA seems to help sometimes but not always. It seems like it starts when I enter an invalid code to try to display an Invalid Code Entered error msg.

Let me know if it works or doesn't work. Thanks

Download the icloud3-v2.2.0-rc11c.zip [here](https://github.com/gcobb321/icloud3/tree/master/v2.2.0-Release%20Candidate).

**Comment by SaturnusDJ** [comment]
Thanks for your work. I have extremely busy times for about 2 weeks, I am not sure if I will be able to try it in between, but after that for sure. You will hear from me.

**Comment by SaturnusDJ** [comment]
It's been too long, I have been lazy...
Yet, I can confirm now it is working. Great work. 👌

**Comment by gcobb321** [comment]
Thanks for your initial comments. They got me to think about 2fa and fmf . And then looking at how fmf and famshr tracking methods were similar yet different and I could make minor, well not so minor, coding changes that provided a lot of options for everyone. 

Note that v2.2 is out, I’m going to be lazy!

**Comment by SaturnusDJ** [comment]
Have a good one. 👍 

I have a questions if that is okay. 😬 

1. For the friend that shares the location via fmf, but has no other connection to HA at all, what would be the best configuration? Since there won't be any active zone triggers etc., will it be fully depended on the device pushing out updates? I think there should be a manual option as well, because you can push fmf app to get an update by force closing it, however the config options `inzone_interval` and `stationary_inzone_interval` do not cover when someone is not in a zone and not stationary, although one would expect the device itself to push out updates regularly then. Yeah what would be the best config for a non HA user?


Feedback:
1. When starting this component, I see my friend's initial location is in the home zone, I do not see that back in the historic state of the device but it gets recorded somewhere so that when adding this to lovelace:
```
    - type: map
      default_zoom: 13
      dark_mode: true
      hours_to_show: 24
      entities:
      - device_tracker.friend_iphone
```
there always is a historical record in the home zone resulting in a line going there. Would be nice if that can be prevented.

2. I cannot find an explanation for `max_interval` at https://gcobb321.github.io/icloud3/#/chapters/2-config-parms. I guess though it can be used in case someone is not stationary, while not being in a zone, while the device is not pushing out updates or so?

3. I think the `log_level` default is not mentioned at https://gcobb321.github.io/icloud3/#/chapters/2-config-parms.

**Comment by gcobb321** [comment]
1. Install the iOS app on your friends phone to create the mobile_app device_tracker entity so you will not get a startup error message and then, in the iOS App on your friends phone, set Location Permissions to never and turn off Zone enter/exit and location triggers. I just update the iCloud3 docs about that [here](https://gcobb321.github.io/icloud3/#/chapters/1-setting-up-iosapp?id=ios-app-settings)
2. That issue wad discussed in one of the release candidates and it is not possible to prevent that on the initial startup of iCloud3. The location entities need to be initialized with some value to get everything up and running. I tried using the iOS App location but that is not actually set up until after iCloud3 gets up and running because of timings and internet communication issues. If you restart iCloud3 from the Event Log > Actions > Restart iCloud3, it does use the last known location instead of Home.
3. I forgot to add max_interval before and have just uploaded a new Configuration Parameters Chapter that describes it [here](https://gcobb321.github.io/icloud3/#/chapters/2-config-parms?id=max_interval).
4. The log_level is discussed at the end of the Configuration Parameters Chapter [here](https://gcobb321.github.io/icloud3/#/chapters/2-config-parms?id=log_level).

**Comment by SaturnusDJ** [comment]
1. What about the case if someone would not want to install that?
2. Should HA itself be changed somewhere to accomplish it? I actually tried restarting from the Lovelace card multiple times, the friend then jumps straight into the home zone and back to real location after a while.
3. Nice, thanks.
4. Yes I found the description but did not found what its default value after installation is.

**Comment by gcobb321** [comment]
I just uploaded iCloud3 v2.2.1a into the development code - v2.2.1 directory [here](https://github.com/gcobb321/icloud3/tree/master/development%20area%20-%20v2.2.1) that supports phones without the ios app. Add ', noiosapp' to the track_devices parameter of the phone without the ios app.

Download the zip file and unzip it into the icloud3 directory.

**Comment by SaturnusDJ** [comment]
Ok got it running, the b version from today, thanks.

What will this new support do? Can it be smart like asking Apple to update more if it sees that the device is not stationary?

Possible bug (also in earlier version): I have `unit_of_measurement: km` yet the number seems not to be converted from miles to km, only km text is added behind it.

**Comment by gcobb321** [comment]
iCloud3 varies the interval based on the distance from home, old location data, poor gps data, battery state several other factors. The interval is displayed in the Event Log and will change when you are not in a zone.

I found a problem with the unit_of_measurement: km config parameter. If it is in the HA configuration.yaml file, the value displayed is correct. If it is in the config_ic3.yaml file, the value displayed is miles with a ‘m’ after it. This has been fixed in v2.2.1c (not updated yet).

Put unit_of_measurement: km in the iCloud3 parameters in the HA configuration.yaml file and see if that displays a correct number.

**Comment by SaturnusDJ** [comment]
Distance was indeed fixed in latest release. 👍

**Comment by SaturnusDJ** [comment]
Something I do not understand: Sometimes when opening the _Find my_ app, the location of a person is updated there, in this case with high precision. Yet, when then triggering a manual update via _Update phone's location_ in HA, that update is not picked up.

**Comment by SaturnusDJ** [comment]
@gcobb321 
I am tracking my friend via my iCloud account. Can I track myself too (sharing location back to that friend) in HA? Or do I then need a 'host' account that receives locations from me and friend?

When adding myself in config_ic3.yaml:
```iCloud3 Error for Me (me) > The email address `x@y.com` is invalid or is not in the FmF contact list and will not be tracked.```

**Comment by gcobb321** [comment]
Add yourself to the FmF contact list in the FmF account you are using to track your friend. Make sure you can see yourself in the FindMy App. 

I track both myself and my wife in my FmF list using my main account without any issues.

**Comment by SaturnusDJ** [comment]
The FmF account is me. When I try to share my location with myself, it refuses. _You cannot share with yourself._
Technically maybe it is still possible to extract my location from my account since I do share it with the friend plus there is the feature for finding back devices etc.

**Comment by SaturnusDJ** [comment]
Seems like since we use iOS 14 the HA addon barely gets get a fresh location anymore if manual update is requested. It's just minutes behind while Find My app is live. Maybe Apple limited something.

**Comment by gcobb321** [comment]
I'm running iOS 14.3, and have been for a while. I haven't noticed any issues related to poor iOS App response time. Do the HA logs indicate any problems or show when the requests were sent vs the update coming back? Does the iOS App Event Log show any delays or errors?

**Comment by SaturnusDJ** [comment]
HA log and iCloud3 Event Log look fine.

```
Results: From-Home > CurrZone-Away, GPS-(X, Y)/Zm), Interval-1 hr, Dist-R km, TravTime- (AwayFrom), NextUpdt-17:21:00, Located-16:14:12 (7 min ago), OldLocThreshold-7.5 min, WillMoveIntoStatZoneAfter-16:29:15 (0m)
```
I think it might have to do with the "OldLocThreshold-7.5 min". Not sure why that is high sometimes (even higher than "5 minute maximum" from documentation) and sometimes as low as 1 minute. Maybe it becomes high after the location has not changed much in a short time (traffic jam or lights)? Then the vehicle moves again and updates will be lagging behind. I am using _old_location_threshold: '2 min'_ in config as test now.

Idea: maybe it makes sense when manually updating to ignore the OldLocThreshold anyways?


Edit:
```
Results: From-Home > CurrZone-Stationary, GPS-(X, Y)/Zm), Interval-15 min, Dist-R km, TravTime- (in_zone), NextUpdt-18:25:15, Located-17:53:42 (16.7 min ago), OldLocThreshold-2 min
```
Result after a manual trigger. Location is hundreds of meters off compared to Find My. After the update on Find My another manual trigger in iCloud3 addon does give the correct location. I don't know what is going wrong, but am quite confident something changed since we went to iOS 14.
